### ISSUE TITLE
SimpleSAMLPHP 2.1.1 language selection fails

### ISSUE THREAD

**Issue by kakopedreros** [simplesamlphp/simplesamlphp#1914]
**Specifics of your environment**
1. Are you acting as SP/IdP/proxy?
2. SimpleSAMLphp: What version are you using?
3. PHP: What version are you using?
4. Platform: unix or Windows?
5. Webserver: Apache/Nginx/ISS?

**Describe the bug**

**To Reproduce**

**Expected behavior**

**Screenshots or logs**
If applicable, add screenshots and/or log files to help explain your problem.

**Additional context**

# **Help us to keep this issue-tracker clean!  For questions or support, please refer to our [mailing lists]

-----------------------

**### Environment**
SimpleSAMLPHP: 2.1.1 as IdP but just working with the Admin module as well.
PHP: 8.2.13 Remi Repo
Platform: Linux Centos 7.9 (Using a dockers image in a MacBook Pro 2015) see attached dockerfile.
Webserver: Apache 2.4.6

[docker-compose.zip](https://github.com/simplesamlphp/simplesamlphp/files/13602991/docker-compose.zip)

![Captura de Pantalla 2023-12-07 a la(s) 14 04 55](https://github.com/simplesamlphp/simplesamlphp/assets/4590300/dad3d5fe-83f1-4515-945a-b4d978bf8731)

### **Bug description**
Trying to select a language, it's set the language cookie, but it not change the content of simplesamlphp, official modules or external. Does not work even changing the default language in the configuration.
It was review file permission, path of the files, proper configuration of PHP dependencies.
Also tested in RockyLinux8 with the same PHP Remi Repo.
SimpleSAMLPHP installed using composer 2.6.5.

### **Screenshots**
Spanish - Español
![Captura de Pantalla 2023-12-07 a la(s) 13 04 26](https://github.com/simplesamlphp/simplesamlphp/assets/4590300/93b18051-05ac-420f-8f87-01269b69fffa)
English (Default)
![Captura de Pantalla 2023-12-07 a la(s) 13 04 19](https://github.com/simplesamlphp/simplesamlphp/assets/4590300/7283442d-489f-411e-bd5a-042cf12c52ad)
(Brazilian Portuguese)
![Captura de Pantalla 2023-12-07 a la(s) 13 04 14](https://github.com/simplesamlphp/simplesamlphp/assets/4590300/4e2a68f4-af27-4dbf-9a1a-e64f47c563b7)

**### Expected Bed**
Show the content of the different modules in the selected language.
It was tested in several browsers in MacOS and Windows 10.

**### LOGS**
```
[Thu Dec 07 15:57:42.129274 2023] [php:notice] [pid 35] [client 192.168.167.0:51221] %date{M j H:i:s} simplesamlphp DEBUG [TRd79624c4] Localization: load domain 'messages' at '/var/www/html/vendor/simplesamlphp/simplesamlphp/locales', referer: http://simplesamlphp.dev.orb.local/public/module.php/core/welcome
[Thu Dec 07 15:57:42.129335 2023] [php:notice] [pid 35] [client 192.168.167.0:51221] %date{M j H:i:s} simplesamlphp DEBUG [TRd79624c4] Trying langpath for 'pt-br' as '/var/www/html/vendor/simplesamlphp/simplesamlphp/locales/pt-br/LC_MESSAGES/', referer: http://simplesamlphp.dev.orb.local/public/module.php/core/welcome
[Thu Dec 07 15:57:42.129343 2023] [php:notice] [pid 35] [client 192.168.167.0:51221] %date{M j H:i:s} simplesamlphp DEBUG [TRd79624c4] /public/module.php/core/welcome - Localization file 'messages.mo' not found or not readable in '/var/www/html/vendor/simplesamlphp/simplesamlphp/locales/pt-br/LC_MESSAGES/', falling back to default, referer: http://simplesamlphp.dev.orb.local/public/module.php/core/welcome
[Thu Dec 07 15:57:42.129348 2023] [php:notice] [pid 35] [client 192.168.167.0:51221] %date{M j H:i:s} simplesamlphp DEBUG [TRd79624c4] Localization: load domain 'core' at '/var/www/html/vendor/simplesamlphp/simplesamlphp/modules/core/locales', referer: http://simplesamlphp.dev.orb.local/public/module.php/core/welcome
[Thu Dec 07 15:57:42.129352 2023] [php:notice] [pid 35] [client 192.168.167.0:51221] %date{M j H:i:s} simplesamlphp DEBUG [TRd79624c4] Trying langpath for 'pt-br' as '/var/www/html/vendor/simplesamlphp/simplesamlphp/modules/core/locales/pt-br/LC_MESSAGES/', referer: http://simplesamlphp.dev.orb.local/public/module.php/core/welcome
[Thu Dec 07 15:57:42.129357 2023] [php:notice] [pid 35] [client 192.168.167.0:51221] %date{M j H:i:s} simplesamlphp DEBUG [TRd79624c4] /public/module.php/core/welcome - Localization file 'core.mo' not found or not readable in '/var/www/html/vendor/simplesamlphp/simplesamlphp/modules/core/locales/pt-br/LC_MESSAGES/', falling back to default, referer: http://simplesamlphp.dev.orb.local/public/module.php/core/welcome
[Thu Dec 07 15:57:42.129368 2023] [php:notice] [pid 35] [client 192.168.167.0:51221] %date{M j H:i:s} simplesamlphp ERR [TRd79624c4] Error creating session: Headers already sent., referer: http://simplesamlphp.dev.orb.local/public/module.php/core/welcome
```

**Comment by tvdijen** [comment]
Looking at this log and seeing stuff like `%date{M j H:i:s}`, it seems you have more issues..

**Comment by tvdijen** [comment]
How did you install? If the .mo-files do not exist, it means something went wrong during installation. You should be able to generate then using `bin/translations translations:update:binary`

**Comment by kakopedreros** [comment]
I just verify, there is not .mo files, only .po files.

I install simplesamlphp using composer.

When i run the script you mentioned, it gives me the following error:

```
[root@sso bin]# ./translations translations:update:binary
PHP Warning:  require(/var/www/html/vendor/simplesamlphp/simplesamlphp/bin/../vendor/autoload.php): Failed to open stream: No such file or directory in /var/www/html/vendor/simplesamlphp/simplesamlphp/bin/translations on line 12
PHP Fatal error:  Uncaught Error: Failed opening required '/var/www/html/vendor/simplesamlphp/simplesamlphp/bin/../vendor/autoload.php' (include_path='.:/usr/share/pear:/usr/share/php') in /var/www/html/vendor/simplesamlphp/simplesamlphp/bin/translations:12
Stack trace:
#0 {main}
  thrown in /var/www/html/vendor/simplesamlphp/simplesamlphp/bin/translations on line 12
```

**Comment by tvdijen** [comment]
OK, I can confirm that when requiring `simplesamlphp/simplesamlphp` from a secondary project, the .mo-files are not generated. Maybe it's best for you to stay on 2.0 for a while, because this is quite a big issue that may not be solved over night

**Comment by monkeyiq** [comment]
It seems that the issue here is that the code behind `bin/translations` assumes there is a ./vendor directory in the root of the simplesamlphp installation. When I make+move to a new directory and run the below I get no `mo` files.
```
composer require simplesamlphp/simplesamlphp
```

This should be done when the script command is executed but fails to find `./bin/../vendor/...` and fails.
```
composer run-script post-install-cmd
```

As a hack I updated the post-install-cmd in composer.json to first ensure the deps are installed and then run the translations script and things worked as expected.
```
        "post-install-cmd": [
            "composer update",
            "php bin/translations translations:update:binary"
        ],
```

It seems I should be using `@composer` in the above. https://getcomposer.org/doc/articles/scripts.md#calling-composer-commands

I am not too sure this is the right way but it seems to be a working way.

**Comment by monkeyiq** [comment]
I have made a cleaned up version of the proposal in https://github.com/simplesamlphp/simplesamlphp/issues/1914#issuecomment-1951475336 into https://github.com/simplesamlphp/simplesamlphp/pull/1967

**Comment by tvdijen** [comment]
The issue should be resolved in v2.2.0
